alf/optimizers/optimizers_test.py

```diff
@@ -252,6 +253,64 @@ def _test(i, sampled_predictive=False):
         self.assertLess(mean_err, 0.5)
         self.assertLess(cov_err, 0.5)
 
+    @parameterized.parameters(
+        dict(capacity_ratio=0.5, masked_out_value=None),
+        dict(capacity_ratio=0.5, masked_out_value=0),
+        dict(
+            capacity_ratio=LinearScheduler(
+                progress_type="percent", schedule=[(0, 0.5), (1., 1)]),
+            masked_out_value=0),
+    )
+    def test_capacity_scheduling(self, capacity_ratio, masked_out_value):
+        layer = torch.nn.Linear(5, 3)
+        x = torch.randn(2, 5)
+        y = layer(x)
+        loss = torch.sum(y**2)
+        clip_norm = 1e-4
+        opt = AdamTF(
+            lr=0.1,
+            gradient_clipping=clip_norm,
+            clip_by_global_norm=True,
+            capacity_ratio=capacity_ratio,
+            masked_out_value=masked_out_value,
+            min_capacity=1)
+        opt.add_param_group({'params': layer.parameters()})
+
+        local_rng = torch.Generator(alf.get_default_device())
+
+        def _generate_rand_matrix_for_capacity_mask(opt):
+            rand_matrix_for_capacity_mask = {}
+            opt_state = opt.state
+            for param_group in opt.param_groups:
+                for p in param_group['params']:
+                    param_state = opt_state[p]
+                    if param_state != {}:
+                        self.assertTrue(
+                            'rng_state' in param_state,
+                            "`rng_state` should be in parameter state")
+
+                        rng_state = param_state['rng_state']
+                        local_rng.set_state(rng_state)
+
+                        # use str type as the key so that later map_structure can be used
+                        # for nested checking
+                        rand_matrix_for_capacity_mask[str(id(p))] = torch.rand(
+                            p.shape, generator=local_rng)
+            return rand_matrix_for_capacity_mask
+
+        rand_matrix0 = _generate_rand_matrix_for_capacity_mask(opt)
+
+        opt.zero_grad()
```

Comment:
Instead of comparing with a mask generated outside of opt. It's better to infer the mask from the parameter before and after the backward() call. And the mask should not change with multiple backward() calls.